src/vma/lwip/tcp_in.c

```diff
@@ -341,19 +341,24 @@ tcp_listen_input(struct tcp_pcb_listen *pcb, tcp_in_data* in_data)
     return ERR_OK;
   }
 
-  if (in_data->flags & TCP_FIN) {
-    /* An incoming FIN should be ignored. Return. */
-    return ERR_OK;
-  }
-
   /* In the LISTEN state, we check for incoming SYN segments,
      creates a new PCB, and responds with a SYN|ACK. */
   if (in_data->flags & TCP_ACK) {
     /* For incoming segments with the ACK flag set, respond with a
        RST. */
+    u16_t flags = TCP_RST;
     LWIP_DEBUGF(TCP_RST_DEBUG, ("tcp_listen_input: ACK in LISTEN, sending reset\n"));
-    tcp_rst(in_data->ackno + 1, in_data->seqno + in_data->tcplen,
-      in_data->tcphdr->dest, in_data->tcphdr->src, NULL);
+    TCP_EVENT_CLONE_PCB(pcb, &npcb, ERR_OK, rc);
+    if (npcb == NULL) {
+      LWIP_DEBUGF(TCP_DEBUG, ("tcp_listen_input: could not allocate PCB\n"));
+      TCP_STATS_INC(tcp.memerr);
+      return ERR_MEM;
+    }
+    ip_addr_copy(npcb->local_ip, in_data->iphdr->dest);
+    ip_addr_copy(npcb->remote_ip, in_data->iphdr->src);
+    npcb->remote_port = in_data->tcphdr->src;
+    TCP_DST_NC_SEND(pcb, npcb, ERR_OK, rc);
+    tcp_rst(in_data->ackno, 0, in_data->tcphdr->dest, in_data->tcphdr->src, npcb, &flags);
```

Comment:
I dont see here any free/cleanup for the newpcb.
Maybe the valgrind is realted to this...

Reply:
Implicit deallocation is used. This code cannot be "traced" with Valgrind using sockperf.